src/codegen/compile.ml

```diff
@@ -540,6 +547,14 @@ module E = struct
     in
     let gc_fn = if !Flags.force_gc then gc_fn else "schedule_" ^ gc_fn in
     call_import env "rts" (gc_fn ^ "_gc")
+
+  (* See Note [Candid subtype checks] *)
+  let add_typtbl_typ (env : t) ty : Int32.t =
+    reg env.typtbl_typs ty
```

Comment:
It seems this may be called multiple times for the same `ty`. Should it look if the same type has been registered before?

Reply:
It should probably do the same thing as 
```
  let func_type (env : t) ty =
    let rec go i = function
      | [] -> env.func_types := !(env.func_types) @ [ ty ]; Int32.of_int i
      | ty'::tys when ty = ty' -> Int32.of_int i
      | _ :: tys -> go (i+1) tys
       in
    go 0 !(env.func_types)
```
Maybe this can even be extracted into a `lookup_or_reg` helper (like `reg`). Not sure if that works well with the different equalities on ocaml types.

Reply:
Or if you don't like the linear scan, add a second field to the env with a `i32 type_map` (or however that's called) to find existing entries.

Reply:
Yeah, I considered this but didn't think it was that important since the size of the type table is not affected and this was just more straigthforward (one entry per static test). But sure, I'll jump.

Reply:
It's kind of ugly to do this and I know realize it's actually unecessary. `idl_sub` and `add_typtblt_typ` will be called at most "once" for a given type because of the way we share the serialization code. 

Though "once" at the moment actually means twice because we generate different code for extended and unextended serialization, but I want to fix that in future (#3185)

Reply:
> since the size of the type table is not affected

But the size of the array with the table indices, right?

> it's actually unnecessary

Ok, if it’s shared anyways, then fine. Maybe leave a comment near this function to explain that.

Reply:
Done